### PR TITLE
Client refactoring

### PR DESCRIPTION
**Pull request by vaage**
Refactoring the Client to be lighter-weight and easier to work with.

### CODE REVIEW

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -14,310 +14,384 @@
...
+    panel.register("u-add", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        final String name = args.hasNext() ? args.nextLine().trim() : "";
+        if (name.length() > 0) {
+          final UserContext user = context.create(name);
+          if (user == null) {
+            System.out.println("ERROR: Failed to create new user");
+          } else {
+            panels.push(createUserPanel(user));
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: Don't sign in as new user. Leave signed-in user unchanged.

File: src/codeu/chat/client/commandline/Chat.java [hunk 2]
```diff
@@ -14,310 +14,384 @@
...
-
-      // TODO: Implement m-next command to jump to an index in the message chain.
-      if (!clientContext.conversation.hasCurrent()) {
-        System.out.println("ERROR: No conversation selected.");
-      } else if (!tokenScanner.hasNextInt()) {
-        System.out.println("Command requires an integer message index.");
-      } else {
-        clientContext.message.selectMessage(tokenScanner.nextInt());
+      // Find the first conversation with the given name and return its context.
+      // If not conversation has the given name, this will return null.
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: s/not/no

File: src/codeu/chat/client/commandline/Chat.java [hunk 3]
```diff
@@ -14,310 +14,384 @@
...
 
-    final Scanner tokenScanner = new Scanner(lineScanner.nextLine());
-    if (!tokenScanner.hasNext()) {
-      return;
-    }
-    final String token = tokenScanner.next();
+    final String command = tokens.hasNext() ? tokens.next() : "";
 
-    if (token.equals("exit")) {
+    // Because "exit" and "back" are appliable to every panel, handle
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: s/appliable/applicable ??

File: src/codeu/chat/client/commandline/Chat.java [hunk 4]
```diff
@@ -14,310 +14,384 @@
...
+    // those commands here to avoid having to implement them for each
+    // panel.
 
-      alive = false;
+    if ("exit".equals(command)) {
+      // The user does not want to process anymore commands
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: anymore should be two words

File: src/codeu/chat/client/commandline/Chat.java [hunk 5]
```diff
@@ -14,310 +14,384 @@
...
-      if (!clientContext.user.hasCurrent()) {
-        System.out.println("ERROR: Not signed in.");
-      } else {
-        if (!tokenScanner.hasNext()) {
-          System.out.println("ERROR: Conversation title not supplied.");
+    });
+
+    // U-SIGN-IN (sign in user)
+    //
+    // Add a command to sign-in as a user when the user enters "u-sign-in"
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: "when the user enters "u-sign-in"" is kind of redundant

File: src/codeu/chat/client/commandline/Chat.java [hunk 6]
```diff
@@ -14,310 +14,384 @@
...
+    //
+    // Add a command that will print a list of all commands and their
+    // descriptions when the user enters "help" while on the user panel.
+    //
+    panel.register("help", new Panel.Command() {
+      @Override
+      public void invoke(Scanner args) {
+        System.out.println("USER MODE");
+        System.out.println("  c-list");
+        System.out.println("    List all conversation that the current user can interact with.");
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: conversation should be plural

File: src/codeu/chat/client/commandline/Chat.java [hunk 7]
```diff
@@ -14,310 +14,384 @@
...
+              conversation.conversation.title,
+              conversation.conversation.id);
+        }
+      }
+    });
+
+    // C-ADD (add conversation)
+    //
+    // Add a command that will create and join a new conversation when the user
+    // enters "c-add" while on the user panel.
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: "when the user enters "c-add" while on the user panel" is not really needed.  The command name is the first thing you see on the next source line.

File: src/codeu/chat/client/commandline/Chat.java [hunk 8]
```diff
@@ -14,310 +14,384 @@
...
 
-      alive = false;
+    if ("exit".equals(command)) {
+      // The user does not want to process anymore commands
+      return false;
+    }
 
-    } else if (token.equals("help")) {
+    // Do not allow the root panel to be removed so make sure there
+    // are enough panels before popping.
```
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: Everything after "Do not allow the root panel to be removed" is kind of obvious.

File: src/codeu/chat/client/commandline/Chat.java (hunk 7 above)
Comment (vaage) [inline src/codeu/chat/client/commandline/Chat.java]: As per a discussion with Riccardo, I am leaving the verbose comment as what may be easy for use to read, may not be for students. Async operations and callbacks are a more advance concept and explaining when the code will be executed allows those who are not comfortable with the concept to get introduced to it in an easier way.

File: src/codeu/chat/client/commandline/Chat.java (hunk 5 above)
Comment (vaage) [inline src/codeu/chat/client/commandline/Chat.java]: As per a discussion with Riccardo, I am leaving the verbose comment as what may be easy for use to read, may not be for students. Async operations and callbacks are a more advance concept and explaining when the code will be executed allows those who are not comfortable with the concept to get introduced to it in an easier way.

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (wgenemorgan) [inline src/codeu/chat/client/commandline/Chat.java]: What is the decision here?

Comment (vaage) [inline src/codeu/chat/client/commandline/Chat.java]: If you look at the latest patch set I removed the sign-in part of u-add.

**Pull request merged**